### PR TITLE
Client/login authguard

### PR DESCRIPTION
**Pull request by maxiemwagenaar**

### CODE REVIEW

File: client/src/app/keycloak-init.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import { KeycloakService } from 'keycloak-angular';
+
+export function initializer(keycloak: KeycloakService): () => Promise<any> {
+  return (): Promise<any> => keycloak.init({
+    config: {
```
Comment (jpastena) [inline client/src/app/keycloak-init.ts]: move the creation of the object outside, and give proper types

File: client/src/app/keycloak-init.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      clientId: 'demo-frontend'
+    },
+    initOptions: {
+      onLoad: 'login-required',
+      checkLoginIframe: false
+    },
+    enableBearerInterceptor: true,
+    bearerExcludedUrls: [
+      '/assets',
+      '/clients/public'
```
Comment (jpastena) [inline client/src/app/keycloak-init.ts]: these endpoints does not exist on our application

File: client/src/app/shared/components/header/header.component.html [hunk 3]
```diff
@@ -7,15 +7,16 @@
...
     </div>
 
     <div class="navbar-nav w-100 justify-content-end">
-      <a class="nav-item nav-link active" [routerLink]="['/account']">
+      <a class="nav-item nav-link active" *ngIf="auth.isAdmin()" [routerLink]="['/account']">
         Account
-        <fa class="icon" size="2x" name="user"></fa>
+        <fa class="icon" size="1x" name="user"></fa>
       </a>
+      <a class="nav-item nav-link active" routerLinkActive="active" (click)="auth.logout()">Logout</a>
```
Comment (jpastena) [inline client/src/app/shared/components/header/header.component.html]: the active class and the active routerlinkactive are in conflict

File: client/src/app/shared/services/authentication/app-auth-guard.service.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import { Injectable } from '@angular/core';
+import { ActivatedRouteSnapshot, Router, RouterStateSnapshot } from '@angular/router';
+import { KeycloakAuthGuard, KeycloakService } from 'keycloak-angular';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AppAuthGuard extends KeycloakAuthGuard {
+  constructor(protected router: Router, protected keycloakAngular: KeycloakService) {
```
Comment (jpastena) [inline client/src/app/shared/services/authentication/app-auth-guard.service.ts]: rename keycloakAngular to keycloakService

File: client/src/app/shared/services/authentication/app-auth-guard.service.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      }
+    });
+  }
+
+  public isAdmin(): boolean {
+    if (this.roles.includes('admin')) {
+      return true;
+    }
+    return false;
+  }
```
Comment (mischaboldy) [inline client/src/app/shared/services/authentication/app-auth-guard.service.ts]: isn't this just `return this.roles.includes('admin')`

File: client/src/app/shared/services/authentication/app-auth-guard.service.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+@Injectable({
+  providedIn: 'root'
+})
+export class AppAuthGuard extends KeycloakAuthGuard {
+  constructor(protected router: Router, protected keycloakAngular: KeycloakService) {
+    super(router, keycloakAngular);
+  }
+
+  isAccessAllowed(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Promise<boolean> {
+    return new Promise((resolve, reject) => {
```
Comment (mischaboldy) [inline client/src/app/shared/services/authentication/app-auth-guard.service.ts]: why does this return a promise?

File: client/package.json [hunk 7]
```diff
@@ -26,6 +26,7 @@
         "core-js": "^2.5.4",
         "font-awesome": "^4.7.0",
         "jquery": "^3.3.1",
+        "keycloak-angular": "^4.0.2",
```
Comment (jpastena) [inline client/package.json]: your mr, forces the rest of the team to have keycloak up and running to use the frontend. In this case:  

- you can force the rest of the team to install keycloak, setup the configuration, and run on theyr machine
- you can create a new folder in the root, called keycloak, and insert all the files necessary to run and setup keycloak in a readme.md file (in this case, you want probably to use the docker version of keycloak)
- you can make the client using keycloak only if an environment variable is set, so using `ng serve` still works, while `ng serve --env=keycloak` needs to have keycloak up running

File: client/src/app/app.module.ts [hunk 8]
```diff
@@ -36,8 +39,15 @@ import { UploadLogListComponent } from './main/upload-log/list/upload-log-list.c
       provide: HTTP_INTERCEPTORS,
       useClass: HeaderInterceptor,
       multi: true
+    },
+    {
+      provide: APP_INITIALIZER,
+      useFactory: initializer,
```
Comment (jpastena) [inline client/src/app/app.module.ts]: can you give a better name for this one, see the other loader

File: client/src/app/keycloak-init.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import { KeycloakOptions, KeycloakService } from 'keycloak-angular';
+import { environment } from '../environments/environment';
+
+export function initializer(keycloak: KeycloakService): () => Promise<any> {
+  if (environment.keycloak) {
+    const initConfig: KeycloakOptions = {
```
Comment (jpastena) [inline client/src/app/keycloak-init.ts]: for the moment is fine to have this here, but in the end you want, probably to move this object (or part of this object) in the environment. so we can "easily change" the url of the keycloak that we want to connect

**Pull request merged**